src/Services/Identity/Identity.Infrastracture/Implementations/UserRepository.cs

```diff
@@ -0,0 +1,116 @@
+﻿using Identity.Domain.Abstractions.Interfaces;
+using Identity.Domain.Models;
+using Identity.Infrastracture.Data;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.EntityFrameworkCore;
+
+namespace Identity.Infrastracture.Implementations;
+
+public class UserRepository : IUserRepository
+{
+    private readonly UserManager<ApplicationUser> _userManager;
+    private readonly ApplicationDbContext _context;
+
+    public UserRepository(UserManager<ApplicationUser> userManager, ApplicationDbContext context)
+    {
+        _userManager = userManager;
+        _context = context;
+    }
+
+    public async Task<string> CreateUserAsync(ApplicationUser model)
+    {
+        model.Id = Guid.NewGuid().ToString();
+        model.ConcurrencyStamp = Guid.NewGuid().ToString();
+        await _userManager.CreateAsync(model);
+        await _userManager.AddPasswordAsync(model, model.PasswordHash!);
+
+        return model.Id;
+    }
+
+    public async Task DeleteUserAsync(string userId)
+    {
+        var user = await _userManager.FindByIdAsync(userId);
+
+        await _userManager.DeleteAsync(user);
+    }
+
+    public async Task<IEnumerable<ApplicationUser>> GetAllUsersAsync()
+    {
+        return await _userManager.Users.ToListAsync();
+    }
+
+    public async Task<ApplicationUser?> GetUserByEmailAsync(string email)
+    {
+        return await _userManager.FindByEmailAsync(email);
+    }
+
+    public async Task<ApplicationUser?> GetUserByUserNameAsync(string username)
+    {
+        return await _userManager.FindByNameAsync(username);
+    }
+
+    public async Task<string> GetUserRoleAsync(string userId)
+    {
+        var user = await _userManager.FindByIdAsync(userId);
+
+        return (await _userManager.GetRolesAsync(user)).Single();
+    }
+
+    public async Task<ApplicationUser?> GetUserByIdAsync(string id)
+    {
+        var user = await _userManager.FindByIdAsync(id);
+
+        return user;
+    }
+
+    public async Task<string> UpdateUserRole(ApplicationUser user, string roleName)
+    {
+        var role = await _context.Roles.SingleAsync(r => r.Name == roleName);
+
+        var userRole = await _context.UserRoles.SingleAsync(ur => ur.UserId == user.Id);
+
+        userRole.RoleId = role.Id;
+
+        _context.UserRoles.Update(userRole);
+
+        await _context.SaveChangesAsync();
```

Comment:
For asynchronous methods, pass CancellationToken and use it. You can pass it from the controller and propagate it further down the call hierarchy